hazelcast/src/main/java/com/hazelcast/cp/internal/RaftGroupMembershipManager.java

```diff
@@ -464,8 +469,93 @@ private void rebalanceLeaderships() {
             Map<RaftEndpoint, CPMember> members = getMembers();
             Collection<CPGroupId> groupIds = getCpGroupIds();
 
-            final int avgGroupsPerMember = groupIds.size() / members.size();
-            final boolean overAvgAllowed = groupIds.size() % members.size() != 0;
+            Map<CPMember, Entry<Integer, Collection<CPGroupId>>> leadershipPriorityMap
+                    = getLeadershipPriorityMap(members.values());
+
+            Map<CPMember, Collection<CPGroupId>> leaderships
+                    = rebalanceLeadershipsByPriority(leadershipPriorityMap, members, groupIds);
+
+            // If there are several CP members with equal priorities, we should rebalance leadership evenly between them
+            if (leaderships.size() > 1) {
+                rebalanceLeadershipsEvenly(leaderships, groupIds, members.values());
+            }
+        }
+
+        private Map<CPMember, Map.Entry<Integer, Collection<CPGroupId>>> getLeadershipPriorityMap(Collection<CPMember> members) {
```

Comment:
can we use some domain-specific classes instead of relying on generics here? `Map<CPMember, Map.Entry<Integer, Collection<CPGroupId>>>` is super un-intuitive what it actually contains

Reply:
Thanks Tomasz, I agree with your comment.
I think there is no sense to create a specific DTO for this case.
The best option will be to use [BiTuple](https://github.com/hazelcast/hazelcast/blob/master/hazelcast/src/main/java/com/hazelcast/internal/util/BiTuple.java), but this class is not serializable. If we make it serializable, we should make the other tuples serializable as well.